File: serialisering/src/test/java/no/nav/syfo/kafka/soknad/deserializer/MultiFunctionDeserializerTest.java
package no.nav.syfo.kafka.soknad.deserializer;

import no.nav.syfo.kafka.interfaces.Soknad;
import org.apache.kafka.common.header.internals.RecordHeader;
import org.apache.kafka.common.header.internals.RecordHeaders;
import org.junit.jupiter.api.Test;

import static java.nio.charset.StandardCharsets.UTF_8;
import static java.util.Collections.*;
import static no.nav.syfo.kafka.KafkaHeaderConstants.MELDINGSTYPE;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class MultiFunctionDeserializerTest {

    private class Testsoknad implements Soknad {
        public String id;

        public Testsoknad(String id) {
            this.id = id;
        }
    }

    @Test
    void testDefault() {
        String s = "id";

        MultiFunctionDeserializer<Testsoknad> deserializer = new MultiFunctionDeserializer<>(emptyMap(),
                bytes -> new Testsoknad(new String(bytes, UTF_8)));

        Testsoknad testsoknad = deserializer.deserialize("topic",
                new RecordHeaders(singletonList(
                        new RecordHeader(MELDINGSTYPE, "spesifikk".getBytes(UTF_8)))), s.getBytes());

        assertThat(testsoknad.id).isEqualTo("id");
    }

    @Test
    void testIngenDefault() {
        String s = "id";

        MultiFunctionDeserializer<Testsoknad> deserializer = new MultiFunctionDeserializer<>(emptyMap());

        assertThatThrownBy(() -> deserializer.deserialize("topic",
                new RecordHeaders(singletonList(
                        new RecordHeader(MELDINGSTYPE, "spesifikk".getBytes(UTF_8)))), s.getBytes()))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("Default deserializer er ikke definert")
                .hasNoCause();
    }

    @Test
    void testFeilIDefault() {
        String s = "id";

        MultiFunctionDeserializer<Testsoknad> deserializer = new MultiFunctionDeserializer<>(emptyMap(),
                bytes -> {
                    throw new IllegalArgumentException();
                });

        assertThatThrownBy(() -> deserializer.deserialize("topic",
                new RecordHeaders(singletonList(
                        new RecordHeader(MELDINGSTYPE, "spesifikk".getBytes(UTF_8)))), s.getBytes()))
                .isInstanceOf(RuntimeException.class)
                .hasMessage("Feil ved konvertering av bytes til søknad")
                .hasCauseInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void testSpesifikk() {
        String s = "id";

        MultiFunctionDeserializer<Testsoknad> deserializer = new MultiFunctionDeserializer<>(singletonMap("spesifikk",
                ((headers, bytes) -> new Testsoknad(new String(bytes, UTF_8)))));

        Testsoknad testsoknad = deserializer.deserialize("topic",
                new RecordHeaders(singletonList(
                        new RecordHeader(MELDINGSTYPE, "spesifikk".getBytes(UTF_8)))), s.getBytes());

        assertThat(testsoknad.id).isEqualTo("id");
    }

    @Test
    void testFeilISpesifikk() {
        String s = "id";

        MultiFunctionDeserializer<Testsoknad> deserializer = new MultiFunctionDeserializer<>(singletonMap("spesifikk",
                ((headers, bytes) -> {
                    throw new IllegalArgumentException();
                })));

        assertThatThrownBy(() -> deserializer.deserialize("topic",
                new RecordHeaders(singletonList(
                        new RecordHeader(MELDINGSTYPE, "spesifikk".getBytes(UTF_8)))), s.getBytes()))
                .isInstanceOf(RuntimeException.class)
                .hasMessage("Feil ved konvertering av bytes til søknad")
                .hasCauseInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void testSpesifikkMatcherIkkeBrukerDefault() {
        String s = "id";

        MultiFunctionDeserializer<Testsoknad> deserializer = new MultiFunctionDeserializer<>(singletonMap("ikkeRiktigFunksjon",
                ((headers, bytes) -> new Testsoknad("ikkeDenne"))),
                bytes -> new Testsoknad(new String(bytes, UTF_8)));

        Testsoknad testsoknad = deserializer.deserialize("topic",
                new RecordHeaders(singletonList(
                        new RecordHeader(MELDINGSTYPE, "spesifikk".getBytes(UTF_8)))), s.getBytes());

        assertThat(testsoknad.id).isEqualTo("id");
    }
}

File: serialisering/src/test/kotlin/no/nav/syfo/kafka/soknad/deserializer/MultiFunctionDeserializerTest.kt
package no.nav.syfo.kafka.soknad.deserializer

import no.nav.syfo.kafka.MELDINGSTYPE
import org.apache.kafka.common.header.Header
import org.apache.kafka.common.header.internals.RecordHeader
import org.apache.kafka.common.header.internals.RecordHeaders
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import java.nio.charset.StandardCharsets.UTF_8
import java.util.Collections.emptyMap

class MultiFunctionDeserializerTest {

    private inner class Testsoknad(var id: String)

    @Test
    fun testDefault() {
        val deserializer = MultiFunctionDeserializer(emptyMap()) { bytes ->
            bytes?.let { Testsoknad(String(it, UTF_8)) }
        }

        val testsoknad = deserializer.deserialize("topic",
                RecordHeaders(listOf<Header>(RecordHeader(MELDINGSTYPE, "spesifikk".toByteArray(UTF_8)))), "id".toByteArray())

        assertThat(testsoknad?.id).isEqualTo("id")
    }

    @Test
    fun testIngenDefault() {
        val deserializer = MultiFunctionDeserializer<Testsoknad>(emptyMap())

        assertThatThrownBy {
            deserializer.deserialize("topic",
                    RecordHeaders(listOf<Header>(RecordHeader(MELDINGSTYPE, "spesifikk".toByteArray(UTF_8)))), "id".toByteArray())
        }
                .isInstanceOf(IllegalArgumentException::class.java)
                .hasMessage("Default mapperfunksjon er ikke definert")
                .hasNoCause()
    }

    @Test
    fun testFeilIDefault() {
        val deserializer = MultiFunctionDeserializer<Testsoknad>(emptyMap()) { throw IllegalArgumentException() }

        assertThatThrownBy {
            deserializer.deserialize("topic",
                    RecordHeaders(listOf<Header>(RecordHeader(MELDINGSTYPE, "spesifikk".toByteArray(UTF_8)))), "id".toByteArray())
        }
                .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun testSpesifikk() {
        val deserializer = MultiFunctionDeserializer<Testsoknad?>(mapOf("spesifikk" to { _, bytes -> bytes?.let { Testsoknad(String(it, UTF_8)) } }))

        val testsoknad = deserializer.deserialize("topic",
                RecordHeaders(listOf<Header>(RecordHeader(MELDINGSTYPE, "spesifikk".toByteArray(UTF_8)))), "id".toByteArray())

        assertThat(testsoknad!!.id).isEqualTo("id")
    }

    @Test
    fun testFeilISpesifikk() {
        val deserializer = MultiFunctionDeserializer<Testsoknad>(mapOf("spesifikk" to { _, _ -> throw IllegalArgumentException() }))

        assertThatThrownBy {
            deserializer.deserialize("topic",
                    RecordHeaders(listOf<Header>(RecordHeader(MELDINGSTYPE, "spesifikk".toByteArray(UTF_8)))), "id".toByteArray())
        }
                .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun testSpesifikkMatcherIkkeBrukerDefault() {
        val deserializer = MultiFunctionDeserializer(mapOf("ikkeRiktigFunksjon" to
                { _, _ -> Testsoknad("ikkeDenne") })
        ) { bytes -> bytes?.let { Testsoknad(String(it, UTF_8)) } }

        val testsoknad = deserializer.deserialize("topic",
                RecordHeaders(listOf<Header>(RecordHeader(MELDINGSTYPE, "spesifikk".toByteArray(UTF_8)))), "id".toByteArray())

        assertThat(testsoknad!!.id).isEqualTo("id")
    }
}
